### PR TITLE
Lskatz patch 1

### PR DESCRIPTION
**Pull request by lskatz**
- [x] This comment contains a description of what is in the pull request.

EToKi is the engine under EnteroBase including the MLST caller.
I have not included the kraken database but have included all other recommended software from their installation documentation.

- [x] Build your own docker image using a Dockerfile
  - [x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [x] Edit main README.md
- [x] Edit Program_Licenses.md
- [x] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Write a GitHub actions workflow
  - [x] Should be located in .github/workflows/ and named test-<tool>.yml (i.e. .github/workflows/test-spades.yml)
  - [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
  - [x] Have successfully run the workflow "Test <program name> image" in your forked repository

- [x] Build your own docker image using a Dockerfile
  - [x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [x] Edit main README.md
- [x] Ensure tool is listed in Program_Licenses.md
- [x] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Update GitHub actions workflow if needed
- [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [x] Have successfully run the workflow "Test <program name> image" in your forked repository

- [x] Build your own docker image using a Dockerfile
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [x] Ensure tool is listed in Program_Licenses.md
- [x] Ensure a simple container-specific README.md exists in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Update GitHub actions workflow if needed
- [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [x] Have successfully run the workflow "Test <program name> image" in your forked repository

- [x] Update relevant GitHub actions workflow files
- [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [x] Have successfully run the workflow "Test <program name> image" in your forked repository

### CODE REVIEW

File: etoki/1.2/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+RUN cd /usr/local/bin/EToKi && python EToKi.py configure --install
+
+RUN find /usr/local/bin/EToKi -name '*.py' -exec sed -i.bak -e '1 i #!/usr/bin/env python\n# ^^^ inserted corrected shebang for this container' {} \;
+
+# WORKDIR sets working directory
+WORKDIR /data
+
+# A second FROM insruction creates a new stage
+# We use `test` for the test image
+#FROM app as test
```
Comment (erinyoung) [inline etoki/1.2/Dockerfile:91]: This needs to be uncommented

Comment (lskatz) [reply etoki/1.2/Dockerfile]: Done!

**Comment by lskatz** [comment]
I _mostly_ passed the linter except that I used `cd` instead of `WORDIR` and have `USER root` and didn't delete apt lists.

**Comment by lskatz** [comment]
Darnit, hang on, I need to edit it so that it installs `usearch`

**Comment by erinyoung** [comment]
I think you're almost out of the woods

I'm impressed that github actions only takes 10 minutes to build and configure this.

I think it would be nice (but not necessary) to replace all the `which` statements with their actual path when possible. 

Also, usearch doesn't seem to be found. Do you think this is an issue?

I copied this from a github action report.

```
#87 [app 47/48] RUN EToKi.py configure --usearch /usr/local/bin/usearch
#87 0.461 2022-05-03 20:24:05.416984	bbduk ("/opt/bbmap/bbduk.sh") is present. 
#87 0.464 2022-05-03 20:24:05.420896	bbmerge ("/opt/bbmap/bbmerge.sh") is present. 
#87 0.471 2022-05-03 20:24:05.427186	blastn ("/ncbi-blast-2.9.0+/bin/blastn") is present. 
#87 0.509 2022-05-03 20:24:05.465070	bowtie2 ("/opt/bowtie2/bowtie2") is present. 
#87 0.552 2022-05-03 20:24:05.508614	bowtie2build ("/opt/bowtie2/bowtie2-build") is present. 
#87 0.555 2022-05-03 20:24:05.511634	bwa ("/EToKi-1.2/externals/bwa") is present. 
#87 0.559 2022-05-03 20:24:05.515544	diamond ("/usr/local/bin/diamond") is present. 
#87 0.561 2022-05-03 20:24:05.517007	fasttree ("/EToKi-1.2/externals/FastTreeMP-DB") is present. 
#87 0.620 2022-05-03 20:24:05.576285	flye ("/flye/bin/flye") is present. 
#87 1.056 2022-05-03 20:24:06.012683	gatk ("/gatk/GenomeAnalysisTK-3.8-1-0-gf15c1c3ef/GenomeAnalysisTK.jar") is present. 
#87 1.077 2022-05-03 20:24:06.033661	kraken2 ("/kraken2/kraken2") is present. 
#87 1.078 2022-05-03 20:24:06.034177	ERROR - lastal ("/EToKi-1.2/externals/lastal") is not present. 
#87 1.078 2022-05-03 20:24:06.034543	ERROR - lastdb ("/EToKi-1.2/externals/lastdb") is not present. 
#87 1.084 2022-05-03 20:24:06.039964	makeblastdb ("/ncbi-blast-2.9.0+/bin/makeblastdb") is present. 
#87 1.139 2022-05-03 20:24:06.095422	megahit ("/megahit/megahit_v1.1.4_LINUX_CPUONLY_x86_64-bin/megahit") is present. 
#87 1.141 2022-05-03 20:24:06.097058	minimap2 ("/EToKi-1.2/externals/minimap2") is present. 
#87 1.142 2022-05-03 20:24:06.098522	mmseqs ("/mmseqs/bin/mmseqs") is present. 
#87 1.143 2022-05-03 20:24:06.099640	pilercr ("/EToKi-1.2/externals/pilercr") is present. 
#87 1.667 2022-05-03 20:24:06.622881	pilon ("/pilon/pilon-1.22.jar") is present. 
#87 1.668 2022-05-03 20:24:06.624263	rapidnj ("/EToKi-1.2/externals/rapidnj") is present. 
#87 1.670 2022-05-03 20:24:06.626398	raxml ("/standard-RAxML/raxmlHPC") is present. 
#87 1.671 2022-05-03 20:24:06.627700	raxml_ng ("/raxml_ng/raxml-ng") is present. 
#87 1.675 2022-05-03 20:24:06.631552	repair ("/opt/bbmap/repair.sh") is present. 
#87 1.676 2022-05-03 20:24:06.632823	samtools ("/lyve-SET/scripts/samtools") is present. 
#87 1.751 2022-05-03 20:24:06.707700	spades ("/spades/bin/spades.py") is present. 
#87 1.753 2022-05-03 20:24:06.709045	trf ("/EToKi-1.2/externals/trf409.linux64") is present. 
#87 1.754 2022-05-03 20:24:06.710205	ERROR - usearch ("/usr/local/bin/usearch") is not present. 
#87 1.754 2022-05-03 20:24:06.710254	WARNING - kraken_database is not present. 
#87 1.754 You can still use EToKi except the parameter "--kraken" in EToKi assemble will not work.
#87 1.754 Alternatively you can download minikraken2 database using --download_krakenDB or pass an pre-installed database into EToKi using --link_krakenDB.
#87 1.756 2022-05-03 20:24:06.712868	Configuration complete.
#87 DONE 1.8s
```

**Comment by lskatz** [comment]
There will be some decreased functionality but hopefully it does not affect the MLST functionality.  I'll get back to you when possible -- compiling it now with `singularity build etoki.sif docker-daemon://etoki:1.2`.  I'll also fix the issues with `which`.

**Comment by lskatz** [comment]
I unfortunately had to edit the line where `usearch` is called so that it fits `vsearch` parameters. One major change is that `vsearch` does not accept amino acid query/ref and so it will search in the nucleotide space instead. The initial results seem to work on the sample 7-gene MLST and so I think it's at least okay to have in the container.

But the good part is that it compiles!  And the test works!

**Review by erinyoung — CHANGES_REQUESTED** [review]
I think it looks great!

In the etoki/1.2/README.md, could you indicate that you had to change the original tool to run with vsearch?

File: etoki/1.2/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+FROM staphb/mlst:${MLST_VER}                AS mlst
+FROM staphb/orthofinder:${ORTHOFINDER_VER}  AS orthofinder
+FROM staphb/cfsan-snp-pipeline:${CFSAN_VER} AS cfsan
+FROM staphb/flye:${FLYE_VER}                AS flye
+FROM staphb/samtools:${SAMTOOLS_VER}        AS samtools
+FROM staphb/raxml:${RAXML_VER}              AS raxml
+
+FROM ubuntu:jammy as app
+
+LABEL base.image="ubuntu:focal"
```
Comment (erinyoung) [inline etoki/1.2/Dockerfile:34]: I just realized that ubuntu:focal should be ubuntu:jammy

**Comment by lskatz** [comment]
I think I covered the suggestions which were very helpful, thanks!

**Comment by kapsakcj** [comment]
Was able to build successfully, gee what a beast of a docker image.

I did notice a quirk during the testing bits near the end. I see that you purposefully installed `spades.py` 3.15.4 via the staphb docker image, but when the etoki image is built, spades 3.15.1 is installed & priority in the PATH.
```
$ docker_run -ti lskatz/etoki:1.2 /bin/bash -c "spades.py --version; which spades.py"
SPAdes genome assembler v3.14.1
/spades/bin/spades.py
```


Any idea how that happened? Did another program install an older version of spades?

The reason I bring it up is because of this traceback during the building of the etoki image. [This was supposedly fixed in spades 3.15.4](https://github.com/ablab/spades/issues/863)
```
Step 112/112 : RUN bash example.bash
 ---> Running in fa152f9a8de9
2022-05-25 02:39:43.892227      Load in 2 read files from 1 libraries
2022-05-25 02:39:47.581476      Obtained 5773534 bases in 32867 reads after Trimming in Lib 0
--pe examples/prep_out_L1_R1.fastq.gz,examples/prep_out_L1_R2.fastq.gz --se examples/prep_out_L1_SE.fastq.gz
2022-05-25 02:39:48.581561      Load in 3 read files from 2 libraries
2022-05-25 02:39:48.680125      Estimated read length: 175.73541761752625
Traceback (most recent call last):
  File "/spades/bin/spades.py", line 643, in <module>
    main(sys.argv)
  File "/spades/bin/spades.py", line 583, in main
    print_params(log, log_filename, command_line, args, cfg)
  File "/spades/bin/spades.py", line 323, in print_params
    print_used_values(cfg, log)
  File "/spades/bin/spades.py", line 113, in print_used_values
    dataset_data = pyyaml.load(open(cfg["dataset"].yaml_filename))
  File "/spades/share/spades/pyyaml3/__init__.py", line 72, in load
    return loader.get_single_data()
  File "/spades/share/spades/pyyaml3/constructor.py", line 37, in get_single_data
    return self.construct_document(node)
  File "/spades/share/spades/pyyaml3/constructor.py", line 46, in construct_document
    for dummy in generator:
  File "/spades/share/spades/pyyaml3/constructor.py", line 398, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/spades/share/spades/pyyaml3/constructor.py", line 204, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/spades/share/spades/pyyaml3/constructor.py", line 126, in construct_mapping
    if not isinstance(key, collections.Hashable):
AttributeError: module 'collections' has no attribute 'Hashable'
```

**Comment by kapsakcj** [comment]
Ohhhh, Etoki is what is installing these dependencies https://github.com/zheminzhou/EToKi#installation

> All 3rd party programs except for usearch can be automatically installed using configure command:
```
python EToKi.py configure --install --download_krakenDB
```

Can lines 136-156 in the dockerfile be removed? Will Etoki run fine as long as all of these tools are in the PATH?

**Comment by kapsakcj** [comment]
Gah, I should have continued reading the Etoki README:

> You can also use pre-installed 3rd party programs in EToKi, by passing their absolute paths into the program using --path. This argument can be specified multiple times in the same command:

Nevermind. Still not sure how spades 3.14.1 got in there

**Comment by kapsakcj** [comment]
ok I'm going to stop looking at code late at night. Running myself in circles here. 

SPAdes 3.14.1 was definitely brought in from the staphb/shovill:1.1.0 docker image. I assumed it was the latest version, but it isn't.

Might recommend upgrading to SPAdes 3.14.5 via the bespoke docker image `staphb/spades:3.15.4` to avoid the traceback error I mentioned previously

**Review by lskatz — COMMENTED** [review]
Request has been addressed

File: etoki/1.2.1/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+## EToKi itself ##
+#RUN wget https://github.com/zheminzhou/EToKi/archive/refs/tags/${ETOKI_VER}.tar.gz && \
+#  ls / && \
+#  tar zxvf ${ETOKI_VER}.tar.gz && \
+#  rm ${ETOKI_VER}.tar.gz
+# TODO checkout a specific tag or hashtag
+RUN git clone https://github.com/lskatz/EToKi.git /EToKi-${ETOKI_VER} && \
+ cd /EToKi-${ETOKI_VER} && \
+ git checkout 1.2.1 && \
```
Comment (erinyoung) [inline etoki/1.2.1/Dockerfile:149]: if it does stay git checkout, this should probably be
```
git checkout ${ETOKI_VER} && \
```

File: etoki/1.2.1/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+WORKDIR /
+
+## EToKi itself ##
+#RUN wget https://github.com/zheminzhou/EToKi/archive/refs/tags/${ETOKI_VER}.tar.gz && \
+#  ls / && \
+#  tar zxvf ${ETOKI_VER}.tar.gz && \
+#  rm ${ETOKI_VER}.tar.gz
+# TODO checkout a specific tag or hashtag
+RUN git clone https://github.com/lskatz/EToKi.git /EToKi-${ETOKI_VER} && \
```
[the remark this answers is not in the document]
Comment (lskatz) [reply etoki/1.2.1/Dockerfile:147]: Ok adding this in too as a wget. Appreciate that.

**Comment by lskatz** [comment]
Looks like tests passed, and I answered your comments.  Thanks for catching all those mistakes or things that raise eyebrows so that it's polished.

File: etoki/1.2.1/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+WORKDIR /
+
+## EToKi itself ##
+#RUN wget https://github.com/zheminzhou/EToKi/archive/refs/tags/${ETOKI_VER}.tar.gz && \
+#  ls / && \
+#  tar zxvf ${ETOKI_VER}.tar.gz && \
+#  rm ${ETOKI_VER}.tar.gz
+# TODO checkout a specific tag or hashtag
+RUN wget https://github.com/lskatz/EToKi/archive/refs/tags/${ETOKI_VER}.tar.gz && \
+ tar zxvf 1.2.1.tar.gz
```
Comment (erinyoung) [inline etoki/1.2.1/Dockerfile:148]: I think this line is mean to read
```
tar zxvf ${ETOKI_VER}.tar.gz
```

File: .github/workflows/test-etoki.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# This caller workflow builds an image to the "test" stage.
```
Comment (erinyoung) [inline .github/workflows/test-etoki.yml:1]: This isn't needed anymore

**Review by erinyoung — CHANGES_REQUESTED** [review]
It's just one small thing in your dockerfile.

Also, we don't need the test-etoki.yml anymore, so it'd be awsome if you could remove that.

File: etoki/1.2.1/Dockerfile (hunk 5 above)
Comment (lskatz) [reply etoki/1.2.1/Dockerfile]: Yep, thanks!

File: .github/workflows/test-etoki.yml (hunk 6 above)
Comment (lskatz) [reply .github/workflows/test-etoki.yml]: Ok I am removing this yml

**Comment by lskatz** [comment]
Thanks for those comments.  I think I addressed those two comments.  Let me know if there is anything else I can do!

**Comment by erinyoung** [comment]
I'm going to merge this and deploy the image to dockerhub. It should be there _soon_

**Pull request merged**